### ISSUE TITLE
GPU idle time with keras multiprocessing

### ISSUE THREAD

**Issue by DavideRutigliano** [keras-team/keras#13508]
Hi all,

This is the first time i'm trying to use 100% of a gpu and i'm not very expert in that.

I've implemented a custom image generator for keras (extending sequence class) for using it with `fit_generator` method with `multiprocessing = True`. Here is the code of the generator:
`class DataGenerator(Sequence):
    
    def __init__(self,
                 input_dir,
                 ids,
                 labels,
                 #dataframe,
                 num_classes,
                 batch_size=64,
                 input_shape=(3, 224, 224),
                 #shuffle=False,
                 resample=False,
                 ratio=1.):
        
        self.input_dir = input_dir
        
        #self.data = dataframe
        self.ids = ids
        self.labels = labels

        self.num_classes = num_classes
        
        self.batch_size = batch_size
        self.input_shape = input_shape
        
        #self.shuffle = shuffle
        self.resample = resample
        self.ratio = ratio        
        
        self.epoch = 0
        self.on_epoch_end()
        
    def on_epoch_end(self):
        #self.dataframe = self.data.copy()
        seed = self.epoch * self.batch_size
        
        #if self.shuffle == True:
        #    self.dataframe = shuffle(self.dataframe, random_state=seed)
        
        if self.resample == True:
            msss = MultilabelStratifiedShuffleSplit(n_splits=1,
                                                    test_size=1-self.ratio,
                                                    random_state=seed)

            for train_index, _ in msss.split(self.ids, self.labels):
                self.ids = np.array(self.ids)[train_index]
                self.labels = np.array(self.labels)[train_index]
                #self.dataframe = self.dataframe.iloc[train_index]
            
        #self.ids = self.dataframe.index.values.tolist()
        #self.labels = self.dataframe.values.tolist()
        self.n = self.__len__()
        self.epoch += 1

    def __len__(self):
        return int(np.ceil(self.ids.shape[0] / self.batch_size))
        
    def __getitem__(self, step):
        current_ids = self.ids[step * self.batch_size : (step + 1) * self.batch_size]
        X, y = self.__generate_batch(current_ids)
        return X, y
    
    def __get_target(self, id):
        idx = np.where(self.ids == id)
        y = self.labels[idx].astype('float32')
        return y
    
    def __generate_batch(self, ids):
        X = np.zeros((self.batch_size, *self.input_shape))
        y = np.zeros((self.batch_size, self.num_classes))
        for idx, id in enumerate(ids):
            X[idx] = resize_normalize(read_image(self.input_dir, id), self.input_shape)
            y[idx] = self.__get_target(id)
        return X, y`

I call the method:
`score = model.fit_generator(train_generator,
                                steps_per_epoch=train_steps,
                                epochs=epochs,
                                validation_data=validation_generator,
                                validation_steps=validation_steps,
                                callbacks=callbacks,
                                use_multiprocessing=True,
                                max_queue_size=90,
                                workers=mp.cpu_count())`
Where cpu_count() = 4.

Even increasing max_queue_size to 250 with all my cpus always ~95-100%, the gpu continues alternating work and idle times. At the moment i'm using the default value (i.e. max_queue_size=10) having the same results.

## Training
During training, everything looks fine and when i train the model the work is correctly splitted between all the cpus, here is the output using `htop` command:
![Screenshot 2019-10-31 at 15 40 24](https://user-images.githubusercontent.com/37795483/67956676-cb206280-fbf4-11e9-9ba6-e528adecb722.png)

The problem is that when i check instead the gpu usage, i see that there's a lot of idle time while i expect >= 80% all time. Here the output with `nvidia-smi dmon`:

![Screenshot 2019-10-31 at 15 57 32](https://user-images.githubusercontent.com/37795483/67958254-2d7a6280-fbf7-11e9-87ba-1032f59432ce.png)

## Validation

Furthermore, during validation the idle time seems to increase and also (all the 4) cpus works, but only at ~50%.

Tensorflow: tf 2.0.0
Keras: 2.2.0-tf

I'm also using mixed-precision setting the policy as explained in TF2.0 docs.

Does anyone have any idea on how to use my gpu 100%?

**Comment by DavideRutigliano** [comment]
## Edit

It seems after the first epoch and then also in successive epochs the total training times reduces (and also cpu usage ~30-40%), while the gpu usage stays above 55% in average (and not goes from peak to peak as in the figure i posted above).

Is this an intended behaviour?

**Comment by bafonso** [comment]
I have found use_multiprocessing=False works better but I still feel my cores are largely not fully used and my GPU idles a lot unless it's a bigger model.
